src/content_scripts/applyColorSchemeJs.js

```diff
@@ -22,109 +51,294 @@ const ADDON_FAKED_WARNING = "matchMedia has been faked/overwritten by add-on web
+    } else {
+        setMediaQueryList.add(mediaQueryList);
+    }
+
+    setMediaQueryLists.add(mediaQueryList);
 
-            return realQuery.removeListener(...args);
-        },
-    };
+    return hook;
 }
 
 /**
- * The actual JS overwrite for window.matchMedia().
+ * Keep track of listener removal
+ * Returns hook function or null if not found
  *
- * @function
- * @param {string} mediaQueryString
- * @returns {MediaQueryList}
- * @see {@link https://developer.mozilla.org/docs/Web/API/Window/matchMedia}
+ * @private
+ * @param {function} func original listener
+ * @param {MediaQueryList} mediaQueryList
+ * @param {boolean} isOnChange
+ * @returns {function} hook
  */
-function matchMediaOverwrite(...args) {
-    let mediaQueryString, requestedMedia;
-    try {
-        const mediaQueryString = args[0];
-        requestedMedia = getColorTypeFromMediaQuery(mediaQueryString);
-    } catch (e) {
-        // ignore errors and run the real browser function at the bottom
-        requestedMedia = null;
+function trackOffListener(func, mediaQueryList, isOnChange) {
+    let entry = wmFuncToEntry.get(func);
+    if (!entry) {
+        return null;
     }
+    let { hook, setMediaQueryList, setMediaQueryListOnChange } = entry;
 
-    switch (requestedMedia) {
-    case COLOR_STATUS.DARK:
-    case COLOR_STATUS.LIGHT:
-    case COLOR_STATUS.NO_PREFERENCE: {
-        const realColorStatus = getSystemMediaStatus();
-        console.info(
-            "Real media query result: ", realColorStatus, ".",
-            "We fake it to appear like: ", fakedColorStatus, ".",
-            ADDON_FAKED_WARNING
-        );
-
-        // if the real status is the same as the one we fake, just go on
-        // or if the whole feature is disabled, obviously
-        if (fakedColorStatus === realColorStatus || fakedColorStatus === COLOR_STATUS.NO_OVERWRITE) {
-            // continue evaluating real result, no need to fake it
-            break;
-        }
-
-        // faking media queries is hard, and we can only return a fake object
-        const fakeResult = fakeMediaQueryResult(mediaQueryString, requestedMedia);
-        return cloneInto(
-            fakeResult,
-            window,
-            {cloneFunctions: true}
-        );
+    if (isOnChange) {
+        setMediaQueryListOnChange.delete(mediaQueryList);
+    } else {
+        setMediaQueryList.delete(mediaQueryList);
     }
+
+    if (setMediaQueryList.size === 0 && setMediaQueryListOnChange.size === 0) {
+        setMediaQueryLists.delete(mediaQueryList);
     }
 
-    // pass to default function, by default
-    return window.matchMedia(...args);
+    return hook;
+}
+
+/**
+ * Check if an object is a MediaQueryList
+ * The correctness of this relies on Firefox's XPCNativeWrapper
+ *
+ * @private
+ * @param {function} obj
+ * @returns {boolean}
+ */
+function checkIsMediaQueryList(obj) {
+    return (Object.prototype.toString.call(obj) === '[object MediaQueryList]');
+}
+
+const skeleton = {
```

Comment:
Could you add a `jsdoc` thing here to describe what the skelecton is?

https://jsdoc.app/tags-property.html